urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/ops/PolygonOp.java

```diff
@@ -0,0 +1,276 @@
+	 */ 
+	public static SimplePolygon toSimplePolygon(Polygon p) throws GeomException{
+		return toSimplePolygon(p,true);
+	}
+	public static SimplePolygon toSimplePolygon(Polygon p, boolean check) throws GeomException{
+		// check if it is simple polygon
+		if((check && isSimple(p)) ){
+			if(p instanceof SimplePolygon)
+				return (SimplePolygon)p;
+			else
+				return new SimplePolygon(p.points());
+		}
+		
+		/*
+		 * Compute the intersection and walk around outside. 
+		 */
+		int n = p.degree();		
+		//compute upper/lower hull;
+		int ll=p.llPos(); //left lower point 
+
+		//Create IntersectSegment for each edge.
+		IntersectSegment[] edges = new IntersectSegment[n];
+		for(int i=0;i<n;i++){
+			edges[i] = new IntersectSegment(Segment.create(p.point((n+ll+i)%n), p.point((n+ll+i+1)%n)));
+		}
+		
+		//create IntersectPoint for each point
+		IntersectPoint left=null;
+		for(int i=0;i<n;i++){
+			IntersectSegment pre = edges[(n+i-1)%n];
+			IntersectSegment pos = edges[i];
+			IntersectPoint pt = new IntersectPoint(p.point((ll+i+n)%n), pre,pos, true);
+			if(i==0)
+				left = pt;
+			pre.add(pt);
+			pos.add(pt);
+		}
+		
+		//compute intersection point
+		for(int i=0; i<n; i++){
+			for(int j=i+2;j<n;j++){
+				if(i==0&&j==n-1){//neighbour
+					continue;
+				}
+				IntersectSegment s1 = edges[i];
+				IntersectSegment s2 = edges[j];
+				GeomObj2 obj = s1.intersectSegment(s2);
+				if(obj instanceof Point){
+					Point point = (Point)obj;
+					//NOTE: if point is appeared before, create method will merge them, s1 and s2 doesn't add an copy because of Set
+					IntersectPoint ipt= new IntersectPoint(point, s1, s2, point.type()==CohoAPR.type);
+					s1.add(ipt);
+					s2.add(ipt);
+				}else if(obj instanceof Segment){
+					//ASSERT must be APR result. Test
+					Segment seg = (Segment)obj;
+					IntersectPoint ipt = new IntersectPoint(seg.left(), s1, s2, true);
+					s1.add(ipt);
+					s2.add(ipt);
+					ipt = new IntersectPoint(seg.right(), s1, s2, true);
+					s1.add(ipt);
+					s2.add(ipt);
+				}else{
+					//empty
+				}				
+			}
+		}
+		
+		//walk like union
+		List<IntersectPoint> points = new LinkedList<IntersectPoint>(); 
+		IntersectPoint point = left;
+		Point pre = Point.create((ScaleNumber)point.x().sub(1), (ScaleNumber)point.x());
+		Segment currSeg = new Segment(pre,point);//scale type
+		IntersectSegment nextSeg ;
+		do{
+			if(points.contains(point))
+				throw new GeomException("Infinite loop found. The polygon has self loop. Debug it. ");
+			points.add(point);
+			nextSeg = point.advance(currSeg,true);
+			point = nextSeg.advance(point);
+			currSeg = nextSeg;
+		}while(!point.equals(left));		
+		return new SimplePolygon(SimplePolygonOp.toScale(points, (ScaleType)p.type()));
+	}
+	
+	/*
+	 * During the intersection, interval points are created, convert to scale here.
+	 */
+	protected static Point[] toScale(IntersectPoint[] points, ScaleType type){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 126-131 into a separate method.